### PR TITLE
CosmosDiagnostics: Adds Serialization APIs

### PR DESCRIPTION
**Pull request by bchong95**
# CosmosDiagnostics: Adds Serialization APIs

## Description

This PR adds the ability for the user to pick how they want Cosmos Diagnostics serialized for whatever is get for their use case. 

For now we have:

* Text
* Json
* Binary

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs [hunk 1]
```diff
@@ -36,12 +38,45 @@ public CosmosTraceDiagnostics(ITrace trace, string userAgent = null)
 
         public override string ToString()
         {
-            return $"User Agent: {this.UserAgent} {Environment.NewLine}{TraceWriter.TraceToText(this.Value)}";
+            return this.ToTextString();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs:41]: The default needs to be JSON as previously explained.

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs [hunk 2]
```diff
@@ -27,5 +27,45 @@ public virtual TimeSpan GetClientElapsedTime()
...
         /// <returns>The string field <see cref="CosmosDiagnostics"/> instance in the Azure CosmosDB database service.</returns>
         public abstract override string ToString();
+
+        /// <summary>
+        /// Gets the human readable text representation of CosmosDiagnostics. 
+        /// This is the representation that is best used for manual debugging and intellisense.
+        /// </summary>
+        /// <remarks>The format of this response is not a hard contract. It can change from SDK version to SDK version.</remarks>
+        /// <returns>The human readable text representation of CosmosDiagnostics. </returns>
+        public virtual string ToTextString()
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:37]: I disagree with exposing different diagnostics string formats. There should be one format that users need with the simple ToString(). We can write tooling around that format to convert it to make it easier for engineers to root cause.  The same formatting will allow automation to automatically root cause issues and avoid needing humans to root cause problems.

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs [hunk 3]
```diff
@@ -27,5 +27,45 @@ public virtual TimeSpan GetClientElapsedTime()
...
+            throw new NotImplementedException($"{nameof(CosmosDiagnostics)}.{nameof(ToTextString)}");
+        }
+
+        /// <summary>
+        /// Gets the JSON representation of CosmosDiagnostics. 
+        /// This is the representation that is best used for automated analysis.
+        /// </summary>
+        /// <remarks>The format of this response is not a hard contract. It can change from SDK version to SDK version.</remarks>
+        /// <returns>The JSON representation of CosmosDiagnostics. </returns>
+        public virtual string ToJsonString()
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:49]: All the other contracts we have exposed are just adding abstract methods without the default implementation. We haven't had any complaints. This way it's the preferred build exception.

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs [hunk 4]
```diff
@@ -27,5 +27,45 @@ public virtual TimeSpan GetClientElapsedTime()
...
+        /// Gets the binary representation of CosmosDiagnostics. 
+        /// This is the representation that is best used for storage in binary enabled logging.
+        /// </summary>
+        /// <remarks>
+        /// The format of this response is not a hard contract. 
+        /// It can change from SDK version to SDK version.
+        /// If this is sent as information in a support ticket the tooling can parse it.
+        /// </remarks>
+        /// <returns>The binary representation of CosmosDiagnostics. </returns>
+        public virtual ReadOnlyMemory<byte> ToBinary()
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:65]: What is the benefit from a public user perspective of this versus String? I mean, both are opaque, so why would they get it as binary if the most common use case is to log the diagnostics in a sink (AppInsights/etc)

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:49]: Shouldn't these by abstract methods? The whole point of having abstract classes is that we can add new methods without breaking changes, right? Why can't se add them as abstract like `ToString`?

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:49]: Technically introducing abstract members is considered a breaking change - so within V3 SDK this one looks better.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs:49]: @FabianMeiswinkel  if that is the case then we need to change the SDK guidelines and do this for all APIs. I don't see a reason for this specific class to be an exception when all other APIs added did not do this.

**Comment by j82w** [comment]
Closing this PR as this directly conflicts with plans to automate diagnostic analysis.  Having JSON as the output allows us to build tools to iterate and analysis the diagnostics. This can not be done inside the SDK:
1. It's add to much over head
2. The logic can not be updated. If a new scenario is detected or if a bug is found in the analysis customer have to upgrade.
3. Doesn't know about regression later found after the release.

**Pull request closed without merging**